zebra-state/src/service/finalized_state/disk_format.rs

```diff
@@ -232,6 +232,75 @@ impl IntoDisk for transparent::OutPoint {
     }
 }
 
+impl IntoDisk for sprout::tree::Root {
+    type Bytes = [u8; 32];
+
+    fn as_bytes(&self) -> Self::Bytes {
+        self.into()
+    }
+}
+
+impl IntoDisk for sapling::tree::Root {
+    type Bytes = [u8; 32];
+
+    fn as_bytes(&self) -> Self::Bytes {
+        self.into()
+    }
+}
+
+impl IntoDisk for orchard::tree::Root {
+    type Bytes = [u8; 32];
+
+    fn as_bytes(&self) -> Self::Bytes {
+        self.into()
+    }
+}
+
+impl IntoDisk for sprout::tree::NoteCommitmentTree {
+    type Bytes = Vec<u8>;
+
+    fn as_bytes(&self) -> Self::Bytes {
+        bincode::serialize(self).expect("serialization to vec doesn't fail")
```

Comment:
So this and the other `(Into|From)Disk` impls re-introduce `bincode` and rely on the `serde` trait impls in `zebra-chain`. @jvff @teor2345 what do you think?

Reply:
Here's my understanding so far:

We have to use `serde`, because `bridgetree::Frontier` doesn't expose its data as raw bytes:
https://github.com/zcash/incrementalmerkletree/blob/2f65f19fa660e86e7d1e5f594bfa60e9e7156b74/src/bridgetree.rs#L303

Since we must use `serde`, we also have to use some formatter. `bincode` seems small and fast.

Here are my suggestions for this PR:

1. Can we document why we're using `serde` and `bincode` in the code? Can we add this usage as an exception in the `serde` removal ticket (#2331)?

2. Should we adjust the `bincode` settings to be stricter (no trailing bytes)? I think this fix is optional, but it will help the tests detect more kinds of failures.
https://docs.rs/bincode/1.3.3/bincode/config/index.html#options-struct-vs-bincode-functions

3. How can we detect format changes? (If the underlying structs change, the derived `serde` formats could change too.)

I think we'll need both Sapling and Orchard byte test vectors, with at least one note in each.

Reply:
See https://github.com/zcash/librustzcash/pull/417/files

Reply:
Added documentation, made bincode stricter and added a test to ensure the format does not change in a4df422eda

However, as @daira pointed out (thanks!), they're adding methods do the crate that will allow custom serialization. Should we do that instead?

Also, should we add a e.g. version byte to make it possible to change the format later?

Reply:
> Added documentation, made bincode stricter and added a test to ensure the format does not change in [a4df422](https://github.com/ZcashFoundation/zebra/commit/a4df422eda1e9b4cafc1cbb53861db4ff4f37d97)
> 
> However, as @daira pointed out (thanks!), they're adding methods do the crate that will allow custom serialization. Should we do that instead?

The code we have right now works, so it seems like that work is optional.

Let's open a follow-up ticket for removing `bincode` and `serde`?
Or add these changes to the list in the existing `serde` ticket?

> Also, should we add a e.g. version byte to make it possible to change the format later?

We already have a version number for the entire state format, it's used as part of the state path.

Since it's a Zebra-specific format, we just increment the version number and start again from genesis when the format changes. (Adding new column families, or changing the format of existing column families.)

Do we need to increment the database version in this PR?

https://github.com/ZcashFoundation/zebra/blob/a7b0a521d51af4af7294892604c34087e0414330/zebra-state/src/constants.rs#L28-L29

https://github.com/ZcashFoundation/zebra/blob/a7b0a521d51af4af7294892604c34087e0414330/zebra-state/src/config.rs#L82-L93

Reply:
Added required changes to the `serde` ticket https://github.com/ZcashFoundation/zebra/issues/2331

Incremented database version in 2ee6da78. (Thanks for the explanation! I've wondered what the `v5` was in the path but I thought it was related to network upgrades)

Reply:
> Incremented database version in [2ee6da7](https://github.com/ZcashFoundation/zebra/commit/2ee6da78143f46696cfc522200be71acb1f37f50). (Thanks for the explanation! I've wondered what the `v5` was in the path but I thought it was related to network upgrades)

Yep, Zebra state format version 5 is just a coincidence (like transaction v5 in NU5!).